### ISSUE TITLE
[datasets] Optimize dataloaders for faster iterations

### ISSUE THREAD

**Issue by fg-mindee** [mindee/doctr#190]
The current training scripts in `references` are way too slow.
Let's discuss here how we can improve them

**Comment by fg-mindee** [comment]
Let's benchmark our current pipeline
```
train_set = FUNSD(train=True, download=True)

    start_ts = time()

    for _ in range(args.epochs):
        for page, target in tqdm(train_set):
            page = tf.image.resize(tf.cast(page, dtype=tf.float32), (1024, 1024), method='bilinear') / 255.

    print(time() - start_ts)
```
yields 25.6 sec

Using `ImageDataGenerator`

```
image_generator = ImageDataGenerator(rescale=1./255)
data_generator = image_generator.flow_from_directory(
        batch_size=4,
        directory=FOLDER,
        shuffle=True,
        target_size=(1024, 1024),
        interpolation='bilinear',
        class_mode='categorical')

tart_ts = time()

    for _ in range(args.epochs):
        ds_iter = iter(data_generator)
        for _ in tqdm(range(len(data_generator))):
            page, _ = next(ds_iter)


    print(time() - start_ts)
```
yields 112sec

Using `tf.data.Dataset`:
```
list_ds = tf.data.Dataset.list_files(folder)
list_ds = list_ds.shuffle(len(list_ds), reshuffle_each_iteration=False)

def process_img(file_path):
        # label = get_label(file_path)
        # load the raw data from the file as a string
        img = tf.io.read_file(file_path)
        img = tf.image.decode_jpeg(img, channels=3)
        return tf.image.resize(img, [1024, 1024], method='bilinear')


AUTOTUNE = tf.data.AUTOTUNE
labeled_ds = list_ds.map(process_img, num_parallel_calls=AUTOTUNE)
start_ts = time()

for _ in range(args.epochs):
        for img in labeled_ds.take(1):
            pass

print(time() - start_ts)
```
yields 1.36s

Using `tf.data.Dataset` while batching:
```
ds = tf.data.Dataset.list_files(folder)
# ds = ds.shuffle(len(ds), reshuffle_each_iteration=False)

def process_img(file_path):
    # label = get_label(file_path)
    # load the raw data from the file as a string
    img = tf.io.read_file(file_path)
    img = tf.image.decode_jpeg(img, channels=3)
    return tf.image.resize(img, [1024, 1024], method='bilinear')


AUTOTUNE = tf.data.AUTOTUNE
ds = ds.map(process_img, num_parallel_calls=AUTOTUNE)
ds = ds.cache()
ds = ds.shuffle(buffer_size=1000)
ds = ds.batch(batch_size)
ds = ds.prefetch(buffer_size=AUTOTUNE)


tot_ts = 0

    for _ in range(args.epochs):
        ds_iter = iter(ds)
        start_ts = time()
        for _ in range(149 // 4):
            next(ds_iter)
        tot_ts += time() - start_ts


    print(time() - start_ts)
```
yields 1.35s

**Comment by charlesmindee** [comment]
Thanks for the benchmark!

**Comment by charlesmindee** [comment]
We should also add a tf.utils.Sequence loader to the benchmark to compare with our solution

**Comment by fg-mindee** [comment]
Alright, I just checked `tf.utils.Sequence` in the exact same conditions with the same data.

With this snippet:
```
class KerasSequence(tf.keras.utils.Sequence):
    def __init__(self, img_folder, batch_size):

        self.root = img_folder
        self.img_list = os.listdir(img_folder)
        self.batch_size = batch_size

    def __len__(self):
        return math.ceil(len(self.img_list) / self.batch_size)

    def process_sample(self, file_path):

        img = tf.io.read_file(os.path.join(self.root, file_path))
        img = tf.image.decode_jpeg(img, channels=3)
        return tf.image.resize(img, [1024, 1024], method='bilinear')

    def __getitem__(self, idx):

        samples = self.img_list[idx * self.batch_size: min((idx + 1) * self.batch_size, len(self.img_list))]
        samples = list(map(self.process_sample, samples))

        return tf.stack(samples)

keras_seq = KerasSequence("/home/fg/Downloads/tmp_data/images", 4)


    start_ts = time()

    for _ in range(args.epochs):
        for x in keras_seq:
            pass


    print(time() - start_ts)
```

I get 38.50sec ... (the fact that there is no parallelization must be quite significant here)

**Comment by charlesmindee** [comment]
Thanks for the benchmark!

**Comment by fg-mindee** [comment]
For reference, the proposed implementation of #198 :
```
train_set = FUNSD(train=True, download=True, input_size=(1024, 1024))
train_loader = DataLoader(train_set, batch_size=4)

tot = 0

for _ in range(args.epochs):
    train_iter = iter(train_loader)
    start_ts = time()
    for pages, targets in train_iter:
        pages /= 255.
    tot += time() - start_ts
print(tot)
```

yields: ~3.45sec

**Comment by fg-mindee** [comment]
Should we remove the "critical" flag and consider this as a long-term issue or close the issue altogether @charlesmindee ?

**Comment by fg-mindee** [comment]
I'll investigate if there are improvement leads with TF2.5 or otherwise close this!

**Comment by charlesmindee** [comment]
Any updates @fg-mindee ? If not maybe we can close this issue.

**Comment by fg-mindee** [comment]
Not yet, I first want to compare with PyTorch dataloaders to check how far we are from optimized DL dataloaders. TF2.5, from what I've seen so far, optimizes a few things, but there is no new interface for that. Meaning that having the flexibility that we do with our current design, seems to require have TFrecords to work :man_shrugging:

**Comment by fg-mindee** [comment]
So, I dug quite deep into this to make sure we get appropriate data loading pipelines :

Here are my findings using a more comprehension transformation pipeline (comparison above are no longer relevant):

**TensorFlow**
- Dataset basic iteration: ~29sec
- DocTR dataloader: ~18sec

**PyTorch before #362**
- Dataset basic iteration: ~77sec
- DocTR dataloader: ~23sec (actually it was 39sec before I added dynamic workers selection...)

**PyTorch after #362**
- Dataset basic iteration: ~26sec
- DocTR dataloader: ~18sec



Here is the script I used for TensorFlow:
```python
import os

os.environ["TF_CPP_MIN_LOG_LEVEL"] = "2"
os.environ['USE_TF'] = '1'
os.environ["CUDA_VISIBLE_DEVICES"]="-1"

from doctr.datasets import FUNSD, DataLoader
from doctr.transforms import Resize, Compose, LambdaTransformation
from time import time

EPOCHS = 10

train_set = FUNSD(
    train=True,
    download=True,
    sample_transforms=Compose([
        LambdaTransformation(lambda x: x / 255),
        Resize((1024, 1024)),
    ])
)

tot = 0

for _ in range(EPOCHS):
    start_ts = time()
    for pages, targets in train_set:
        pass
    tot += time() - start_ts

print('DS', tot)

train_loader = DataLoader(
    train_set,
    shuffle=False,
    batch_size=4,
    drop_last=False,
    workers=4,
)

_ = next(iter(train_loader))

tot = 0

for _ in range(EPOCHS):
    train_iter = iter(train_loader)
    start_ts = time()
    for pages, targets in train_iter:
        pass
    tot += time() - start_ts

print('DataLoader', tot)
```

and for PyTorch (commented lines were used to get the "before" results)
```python
import os

os.environ['USE_TORCH'] = '1'

from doctr.datasets import FUNSD
from doctr.transforms import Resize
from torch.utils.data import DataLoader, SequentialSampler
from torchvision.transforms import Compose, Lambda, Resize as TResize, ToTensor
from time import time
import multiprocessing as mp

WORKERS = 4
EPOCHS = 10

ds = FUNSD(
    train=True,
    download=True,
    # sample_transforms=Compose([Lambda(lambda x: x/255), Resize((1024, 1024))]),
    sample_transforms=Resize((1024, 1024)),
)

tot = 0

for _ in range(EPOCHS):
    start_ts = time()
    for pages, targets in ds:
        pass
    tot += time() - start_ts

print('DS', tot)

train_loader = DataLoader(
    ds,
    batch_size=4,
    drop_last=False,
    # num_workers=4,
    num_workers=min(16, mp.cpu_count()),
    sampler=SequentialSampler(ds),
    collate_fn=ds.collate_fn,
)

_ = next(iter(train_loader))

tot = 0

for _ in range(EPOCHS):
    train_iter = iter(train_loader)
    start_ts = time()
    for pages, targets in train_iter:
        pass
    tot += time() - start_ts

print("Loader", tot)
```
